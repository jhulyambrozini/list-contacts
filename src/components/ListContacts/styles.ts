import styled from 'styled-components'

export const ListContactsContainer = styled.ul`
  display: flex;
  flex-direction: column;
  margin-bottom: 4.8rem;
  gap: 24px;
  max-height: 100vh;

  overflow-y: scroll;

  li {
    width: 100%;
    list-style: none;
  }

  @media screen and (max-width: 426px) {
    margin: 3.2rem 1.6rem;
  }
`
