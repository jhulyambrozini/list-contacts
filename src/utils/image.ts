const getRandomColor = (): string => {
  const red = Math.floor(Math.random() * 256)
  const green = Math.floor(Math.random() * 256)
  const blue = Math.floor(Math.random() * 256)

  return `rgb(${red},${green},${blue})`
}

export const generateRandomImage = (
  text: string,
  width: number,
  height: number
): string => {
  const canvas = document.createElement('canvas')
  canvas.width = width
  canvas.height = height

  const ctx = canvas.getContext('2d')
  if (!ctx) {
    throw new Error('Canvas 2D context is not supported')
  }

  // Preencher o canvas com uma cor aleatória
  ctx.fillStyle = getRandomColor()
  ctx.fillRect(0, 0, width, height)

  // Configurar o texto no centro
  ctx.fillStyle = 'white'
  ctx.font = '50px Arial'
  ctx.textAlign = 'center'
  ctx.textBaseline = 'middle'

  // Capitalizar a primeira letra da string e colocá-la no centro
  const firstLetter = text.charAt(0).toUpperCase()
  ctx.fillText(firstLetter, width / 2, height / 2)

  // Converter o canvas para um formato de dados URL
  const dataURL = canvas.toDataURL('image/png')
  return dataURL
}

export const validateImageFormat = (file: File): boolean => {
  const allowedFormats = ['image/png', 'image/jpeg']
  return allowedFormats.includes(file.type)
}
