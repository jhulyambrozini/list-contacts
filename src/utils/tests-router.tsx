import { ReactNode } from 'react'
import { BrowserRouter, useNavigate } from 'react-router-dom'
import { render } from '@testing-library/react'

const renderWithRouter = (component: ReactNode) => {
  let navigate: any
  const Wrapper = () => {
    navigate = useNavigate()
    return component
  }

  return {
    ...render(
      <BrowserRouter>
        <Wrapper />
      </BrowserRouter>
    ),
    navigate
  }
}

export default renderWithRouter
