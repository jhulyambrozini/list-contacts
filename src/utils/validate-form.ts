import { FormikProps } from 'formik'
import { ValuesForm } from '../types/Form'

export const getFieldStatus = (
  field: string,
  form: FormikProps<ValuesForm>
) => {
  const isTouched = field in form.touched
  const isInvalid = field in form.errors
  return { isTouched, isInvalid }
}

export const getMessageError = (
  field: string,
  form: FormikProps<ValuesForm>,
  message?: string
) => {
  const { isTouched, isInvalid } = getFieldStatus(field, form)
  return isTouched && isInvalid ? message : false
}

export const getSuccess = (field: string, form: FormikProps<ValuesForm>) => {
  const { isTouched, isInvalid } = getFieldStatus(field, form)
  return isTouched && !isInvalid
}
