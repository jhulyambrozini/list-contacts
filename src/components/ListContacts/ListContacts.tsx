import { useSelector } from "react-redux"

import { Container, Title } from "../../styles"
import { ListContactStyles } from "./styles"

import Contact from "../Contact/Contact"
import { RootReducer } from "../../store"

const ListContacts = () => {
  const {items} = useSelector((state: RootReducer) => state.contact)
  return (
    <Container>
      <Title as='h3'>Lista de contatos</Title>
      <ListContactStyles>
      {items.map(c => (
        <li key={c.id}>
          <Contact name={c.name} email={c.email} tel={c.tel} id={c.id}/>
        </li>
      ))}
      </ListContactStyles>
    </Container>
  )
}

export default ListContacts
