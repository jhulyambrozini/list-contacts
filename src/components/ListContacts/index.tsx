import Contact from '../Contact'
import { ListContactsContainer } from './styles'

const ListContacts = () => {
  return (
    <ListContactsContainer>
      <li>
        <Contact
          email="exemple@gmail.com"
          id={11}
          image="https://placehold.co/90x90/030020/png"
          name="Kamilly Silva"
          tel="99 99999-9999"
        />
      </li>

      <li>
        <Contact
          email="exemple@gmail.com"
          id={11}
          image="https://placehold.co/90x90/030020/png"
          name="Kamilly Silva"
          tel="99 99999-9999"
        />
      </li>
      <li>
        <Contact
          email="exemple@gmail.com"
          id={11}
          image="https://placehold.co/90x90/030020/png"
          name="Kamilly Silva"
          tel="99 99999-9999"
        />
      </li>
      <li>
        <Contact
          email="exemple@gmail.com"
          id={11}
          image="https://placehold.co/90x90/030020/png"
          name="Kamilly Silva"
          tel="99 99999-9999"
        />
      </li>
      <li>
        <Contact
          email="exemple@gmail.co"
          id={11}
          image="https://placehold.co/90x90/030020/png"
          name="Kamilly Silva"
          tel="99 99999-9999"
        />
      </li>
      <li>
        <Contact
          email="exemple@gmail.com"
          id={11}
          image="https://placehold.co/90x90/030020/png"
          name="Kamilly Silva Santos carmello"
          tel="99 99999-9999"
        />
      </li>
    </ListContactsContainer>
  )
}

export default ListContacts
