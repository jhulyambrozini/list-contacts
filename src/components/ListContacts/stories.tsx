import { Meta } from '@storybook/react'
import { Provider } from 'react-redux'

import ListContacts from '.'

import mockStore from '../../mocks/redux-store'

const meta = {
  title: 'components/ListContacts',
  component: ListContacts,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof ListContacts>

export default meta

export const Empty = () => {
  const store = mockStore({
    contact: {
      items: []
    }
  })
  return (
    <Provider store={store}>
      <ListContacts />
    </Provider>
  )
}

export const WithListItems = () => {
  const store = mockStore({
    contact: {
      items: [
        {
          id: '1',
          firstName: 'John',
          lastName: 'Doe',
          email: 'john@example.com',
          tel: '9999-99999',
          image: 'https://placehold.co/90x90/030020/png'
        },
        {
          id: '2',
          firstName: 'Luiza',
          lastName: 'Doe',
          email: 'Luiza@example.com',
          tel: '9999-99999',
          image: 'https://placehold.co/90x90/030020/png'
        },
        {
          id: '3',
          firstName: 'Antonia',
          lastName: 'Doe',
          email: 'Antonia@example.com',
          tel: '9999-99999',
          image: 'https://placehold.co/90x90/030020/png'
        }
      ]
    }
  })

  return (
    <Provider store={store}>
      <ListContacts />
    </Provider>
  )
}
