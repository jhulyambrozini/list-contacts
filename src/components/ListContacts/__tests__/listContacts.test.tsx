import '@testing-library/jest-dom'
import { screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import ListContacts from '..'
import { renderWithProvider } from '../../../utils/tests-redux'

describe('<ListContacts />', () => {
  it('deve renderizar um paragrafo se a lista estiver vazia', () => {
    const { container } = renderWithProvider(
      <MemoryRouter>
        <ListContacts />
      </MemoryRouter>
    )
    expect(
      screen.getByText(/sua agenda estÃ¡ vazia, adicione novos contatos!/i)
    ).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('deve renderizar o iem na lista quando tiver um', async () => {
    const { container } = renderWithProvider(
      <MemoryRouter>
        <ListContacts />
      </MemoryRouter>,
      {
        preloadedState: {
          contact: {
            items: [
              {
                email: 'exemple@gmail.com',
                id: '1',
                image: '/img.png',
                firstName: 'Kamilly',
                lastName: 'Silva',
                tel: '11 99999-9999'
              },
              {
                email: 'exemple@gmail.co',
                id: '2',
                image: '/img.png',
                firstName: 'Alis',
                lastName: 'Silva',
                tel: '11 99999-8888'
              }
            ]
          }
        }
      }
    )

    expect(
      screen.getByRole('heading', { name: /kamilly silva/i })
    ).toBeInTheDocument()
    expect(
      screen.getByRole('heading', { name: /alis silva/i })
    ).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
})
