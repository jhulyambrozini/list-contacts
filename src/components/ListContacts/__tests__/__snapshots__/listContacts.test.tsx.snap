// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ListContacts /> > should render a paragraph if the list is empty 1`] = `
<p
  class="sc-ftLKQv dxDskM"
>
  Sua agenda est√° vazia, adicione novos contatos!
</p>
`;

exports[`<ListContacts /> > should render the item in the list when it has one in the store 1`] = `
<ul
  class="sc-iBAaJG hLsAyM"
>
  <li>
    <div
      class="sc-eDnVMP dCSdhh"
    >
      <div
        class="sc-jSwlEQ bOxcWW"
      >
        <div
          class="line-deco"
        />
        <img
          alt="Kamilly Silva"
          src="/img.png"
        />
      </div>
      <div
        aria-label="contact info"
        class="sc-gKHVLF bSyYha"
      >
        <h2>
          Kamilly Silva
        </h2>
        <span>
          11 99999-9999
        </span>
        <span>
          exemple@gmail.com
        </span>
      </div>
      <button
        class="sc-bcPKhP sc-hLclGa kjNJwx bSMmYf"
        padding="1.6rem"
        title="Editar contato"
        type="button"
      >
        <img
          alt="Icone de editar"
          src="/src/assets/edit-icon.svg"
        />
      </button>
    </div>
  </li>
  <li>
    <div
      class="sc-eDnVMP dCSdhh"
    >
      <div
        class="sc-jSwlEQ bOxcWW"
      >
        <div
          class="line-deco"
        />
        <img
          alt="Alis Silva"
          src="/img.png"
        />
      </div>
      <div
        aria-label="contact info"
        class="sc-gKHVLF bSyYha"
      >
        <h2>
          Alis Silva
        </h2>
        <span>
          11 99999-8888
        </span>
        <span>
          exemple@gmail.co
        </span>
      </div>
      <button
        class="sc-bcPKhP sc-hLclGa kjNJwx bSMmYf"
        padding="1.6rem"
        title="Editar contato"
        type="button"
      >
        <img
          alt="Icone de editar"
          src="/src/assets/edit-icon.svg"
        />
      </button>
    </div>
  </li>
</ul>
`;
